File: src/main/java/cn/thens/logdog/Logdog.java
package cn.thens.logdog;

@SuppressWarnings("SpellCheckingInspection")
@PrettyLogger.Ignored
public final class Logdog {
    public static final int ASSERT = 7;
    public static final int DEBUG = 3;
    public static final int ERROR = 6;
    public static final int INFO = 4;
    public static final int VERBOSE = 2;
    public static final int WARN = 5;

    private static final String DEFAULT_TAG = "Logdog";
    public static Logdog X = new Logdog(DEBUG, DEFAULT_TAG, new PrettyLogger(Loggers.logcat()));
    private static final Logdog EMPTY = new Logdog(DEBUG, DEFAULT_TAG, Loggers.empty());

    private final String tag;
    private final int priority;
    private final Logger logger;

    private Logdog(int priority, String tag, Logger logger) {
        this.priority = priority;
        this.tag = tag;
        this.logger = logger;
    }

    public Logdog priority(int priority) {
        return new Logdog(priority, tag, logger);
    }

    public Logdog tag(String tag) {
        return new Logdog(priority, tag, logger);
    }

    public Logdog logger(Logger logger) {
        return new Logdog(priority, tag, logger);
    }

    public Logdog onlyIf(boolean sure) {
        return sure ? this : EMPTY;
    }

    public Logdog verbose() {
        return priority(VERBOSE);
    }

    public Logdog debug() {
        return priority(DEBUG);
    }

    public Logdog info() {
        return priority(INFO);
    }

    public Logdog warn() {
        return priority(WARN);
    }

    public Logdog error() {
        return priority(ERROR);
    }

    public Logdog wtf() {
        return priority(ASSERT);
    }

    public Logdog log(Object msg) {
        logger.log(priority, tag, msg);
        return this;
    }
}

File: src/main/java/cn/thens/logdog/Logdog.kt
package cn.thens.logdog

interface Logdog : PrettyLogger.Ignored {
    val isLoggable: Boolean

    fun priority(priority: LogPriority): Logdog

    fun tag(tag: String): Logdog

    fun filter(filter: LogFilter): Logdog

    fun logger(logger: Logger): Logdog

    fun log(msg: Any?): Logdog

    val verbose get() = priority(LogPriority.VERBOSE)

    val debug get() = priority(LogPriority.DEBUG)

    val info get() = priority(LogPriority.INFO)

    val warn get() = priority(LogPriority.WARN)

    val error get() = priority(LogPriority.ERROR)

    val wtf get() = priority(LogPriority.ASSERT)

    val dummy get() = filter(LogFilter.NONE)

    fun require(value: Boolean): Logdog {
        return if (value) dummy else error
    }

    @Deprecated("", ReplaceWith("filter(filter)"))
    fun strategy(filter: LogFilter): Logdog = filter(filter)

    @Deprecated("", ReplaceWith("require(value)"))
    fun requires(value: Boolean): Logdog = require(value)

    companion object X : Logdog by Impl()

    private class Impl(
        private val priority: LogPriority = LogPriority.DEBUG,
        private val tag: String = "Logdog",
        private val filter: LogFilter = LogFilter.ALL,
        private val logger: Logger = PrettyLogger(Logger.logcat()),
    ) : Logdog {

        override val isLoggable: Boolean = filter.isLoggable(priority, tag)

        override fun priority(priority: LogPriority): Logdog {
            return Impl(priority, tag, filter, logger)
        }

        override fun tag(tag: String): Logdog {
            return Impl(priority, tag, filter, logger)
        }

        override fun filter(filter: LogFilter): Logdog {
            return Impl(priority, tag, filter, logger)
        }

        override fun logger(logger: Logger): Logdog {
            return Impl(priority, tag, filter, logger)
        }

        override fun log(msg: Any?): Logdog {
            if (isLoggable) {
                logger.log(priority, tag, msg)
            }
            return this
        }
    }
}